tests/match.d

```diff
@@ -59,3 +59,23 @@ unittest {
 
 	assert(val == 1);
 }
+
+@("Some of some is present")
+@safe @nogc unittest {
+	auto a = some(some(5));
+	const result = a.match!(
+		b => b + 1,
+		() => 0
+	);
+	assert(result == 6);
+}
```

Comment:
I'm not sure about this one. This will result in `result` being an `int`. Shouldn't It  be an `Optional`.

The none-handler should result in an optional as the some-handler in this case:
```
const result = a.match!(
  b => b + 1,
  () => some(0)
);
```

Or the some handler should unwrap the optional it gets to return an int like the none handler.

Reply:
The reason I wrote it this way is because I have some code where I have optionals within optionals. Having to unwrap the inner optional within a match result in this ugly code:
```d
a.match!(
    (inner)
    {
        inner.match!(
            b => doSomething(b);
            () {}
        );
    },
    () {}
);
```
By having `match` unwrap the outer optional automatically, that code is simplified to:
```d
a.match!(
    b => doSomething(b);
    () {}
);
```
If the outer optional needs special treatment, it's much cleaner to put it into either a separate variable or function in my opinion.

That said, I'll admit I haven't thought much about whether the result of something like this should be an optional or not. But, given `match` doesn't wrap the result of a non-nested optional in an optional, I wonder why it should for a nested optional.

If you feel like an alternative solution would be better, I'd be happy to change it. Perhaps an `unwrap` function can be added to which can convert `Optional!(Optional!T)` to `Optional!T` instead of doing that within the `match` function directly?